### PR TITLE
Peloton to Garmin Version 2

### PR DESCRIPTION
**Pull request by philosowaffle**
This is stable enough to be a beta release at this point.  You can test it out by using this [Docker Compose File](https://github.com/philosowaffle/peloton-to-garmin/blob/v2/docker-compose-fit.yaml).  Sample [configuration.local.json](https://github.com/philosowaffle/peloton-to-garmin/blob/v2/PelotonToFit/configuration.example.json).

---

Garmin provides an SDK for writing FIT files: https://developer.garmin.com/fit/overview/

C# SDK is provided, and the code will be just complex enough that I will happily switch back to my native programming language. Python is great and all, but not my most fluent language.

With .net core, should be able to maintain cross platform compatibility in the event I can figure out the FIT protocol.

Pros of FIT:
- Supports way more data than TCX
- I get to switch to C# and stop writing python :D

Cons of FIT:
- Complex protocol
- No python library I could find for writing files
- Will need to rewrite most of the existing code in C#

---
Some notes on why I'm continuing down the c# path and not working with one of the python libraries:

1. https://github.com/jaroslawhartman/withings-sync - Looks great, but I'd need to copy some of their code into this code base and I don't actually understand what the code is doing which means I would not be able to enhance or modify it going forward.
2. https://github.com/dtcooper/python-fitparse - Only supports parsing fit files, not writing them. There is a branch where someone has added writing support but same as point 1, I don't understand it well enough to pull into this project and support long term.
3. The FIT protocol is very complex and the C# library I'm using is easy to use and officially published by Garmin itself.
4. In the age of .netcore, this project should continue to run on multiple OS's.
5. My day job is C#, I only dabble in python on the weekends.

### CODE REVIEW

**Comment by RicardoNiepel** [comment]
Can you please test/ensure, that the fix of PR #100 is also possible and how with FIT?

We should ensure that counting towards Garmin challenges and badges is still possible with FIT file format :)

**Comment by philosowaffle** [comment]
@RicardoNiepel Currently a default DeviceInfo and Author is already written to the FIT files.  So it should be possible for me to add support for the user to customize those values if they choose.  I will probably need help testing and verifying this from the community as badges and challenges are not features I use on Garmin connect.

**Comment by superm1** [comment]
Just FYI - I think normally you don't want to be committing binaries (the stuff in `/dist`).  I would think you're better off separately uploading and artifact somewhere and pulling from that.  For example throw it into a docker registry that is pulled down and only used for making binary releases or something.

File: CONTRIBUTING.md [hunk 1]
```diff
@@ -2,7 +2,28 @@
...
 Enhancements and fixes are always welcome. Feel free to contribute to any of the Issues not already assigned to another person.
 
-* `pip install -r dev_requirements.txt`
-* `cd tests`
-* `pytest`
-* New code should have unit tests coverage
+1. Install [Python 3](https://www.python.org/downloads/) and pip
+1. Install [dotent 5.0 runtime](https://dotnet.microsoft.com/download/dotnet/5.0/runtime)
+1. Clone this repository
+1. In the repository modify the `configuration.local.json`
```
Comment (superm1) [inline CONTRIBUTING.md:8]: in the repo it's got `configuration.example.json`, presumably you mean to rename it and modify it?

File: CONTRIBUTING.md [hunk 2]
```diff
@@ -2,7 +2,28 @@
...
+1. Open the command line
+
+```
+> cd peloton-to-garmin
+> cd python
+> pip install -r requirements.txt
+> cd ../
+> dotnet restore
+> dotnet build
+> dotnet run ./PelotonToGarminConsole/PelotonToGarminConsole.csproj
```
Comment (superm1) [inline CONTRIBUTING.md:19]: Have you moved the directories a bit?  I had to launch it like this:
```
$ dotnet run --project ./src/PelotonToGarminConsole/PelotonToGarminConsole.csproj
```

And that only worked once I added my configuration to this directory:
```
cp configuration.example.json src/PelotonToGarminConsole/
```

**Comment by philosowaffle** [comment]
> Just FYI - I think normally you don't want to be committing binaries (the stuff in /dist). I would think you're better off separately uploading and artifact somewhere and pulling from that. For example throw it into a docker registry that is pulled down and only used for making binary releases or something.

This is intentional since there are many non-code savvy users of this project.  The compiled executables and compiled python projects are there to aide users who are not comfortable touching the command line.  For everyone else, [docker is the preferred installation method](https://github.com/philosowaffle/peloton-to-garmin/wiki/Quick-Start#docker).

File: CONTRIBUTING.md (hunk 2 above)
Comment (philosowaffle) [reply CONTRIBUTING.md:19]: Contributing.md seems to be a little out of date, I'll update it in the next commit. I think the [wiki page](https://github.com/philosowaffle/peloton-to-garmin/wiki/Quick-Start#source) documentation is up to date though.

Comment (philosowaffle) [reply CONTRIBUTING.md:19]: Or, based on the CI failures, I might have messed up something in the rebase, will take a closer look.

**Comment by superm1** [comment]
> > Just FYI - I think normally you don't want to be committing binaries (the stuff in /dist). I would think you're better off separately uploading and artifact somewhere and pulling from that. For example throw it into a docker registry that is pulled down and only used for making binary releases or something.
> 
> This is intentional since there are many non-code savvy users of this project. The compiled executables and compiled python projects are there to aide users who are not comfortable touching the command line. For everyone else, [docker is the preferred installation method](https://github.com/philosowaffle/peloton-to-garmin/wiki/Quick-Start#docker).

Right, makes sense to make it nice and easy for Windows/OSX folks, but still shouldn't you just create artifacts for WIndows and OSX instead of commiting them to the branch?  Maybe this is just because you're pre-release right now for v2, just to make it easy for people until you merge.

File: CONTRIBUTING.md (hunk 2 above)
Comment (superm1) [reply CONTRIBUTING.md:19]: Yeah I saw the CI rereferring to different paths too, hence my confusion.  I think you should just modify the contributing.md to point people at the wiki page for info.

**Comment by philosowaffle** [comment]
> Right, makes sense to make it nice and easy for Windows/OSX folks, but still shouldn't you just create artifacts for WIndows and OSX instead of commiting them to the branch? Maybe this is just because you're pre-release right now for v2, just to make it easy for people until you merge.

By keeping the artifacts with the code, any user can easily grab the latest version without having to wait for an official release, and they can roll back to any given commit. These artifacts are generated by a CI job after every commit, so they are automated and always reflect latest.  At release time (which hasn't happened for v2 yet), the current release artifacts are uploaded to the [Github Release](https://github.com/philosowaffle/peloton-to-garmin/releases).  And preserved there.

I could set up something with Release Candidates, but that just felt like overkill and might complicate the Release page for those not familiar with Github.  The only real downside I see to committing the artifacts is it increases the repo size.

File: CONTRIBUTING.md (hunk 2 above)
Comment (philosowaffle) [reply CONTRIBUTING.md]: Yup, looks like I busted the PR CI job when I rebased. Will need to fix it after work.

**Comment by superm1** [comment]
>These artifacts are generated by a CI job after every commit, so they are automated and always reflect latest. At release time (which hasn't happened for v2 yet), the current release artifacts are uploaded to the Github Release. And preserved there.

I think what you're looking for is to just guide people where to get the artifacts from CI, rather than committing them back to the branch.  If you just guide people like this does: https://docs.github.com/en/actions/managing-workflow-runs/downloading-workflow-artifacts then they can get artifacts for anything pre-release, and as you say when you merge/tag v2 there will be new artifacts on the Releases section.

> The only real downside I see to committing the artifacts is it increases the repo size.

Yeah, so it slows down checkout unless you put the artifacts into Git-LFS.  If it goes on for a while, CI will slow down too.  So IMO it's good to come up with a way to avoid that sooner than later.

**Comment by philosowaffle** [comment]
> I think what you're looking for is to just guide people where to get the artifacts from CI, rather than committing them back to the branch. If you just guide people like this does: https://docs.github.com/en/actions/managing-workflow-runs/downloading-workflow-artifacts then they can get artifacts for anything pre-release, and as you say when you merge/tag v2 there will be new artifacts on the Releases section.

This is good info. I wasn't aware that I could attach artifacts to a GH Action.  I'll investigate further.

**Pull request merged**